src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2Extensions.cs

```diff
@@ -70,102 +68,108 @@ private unsafe static long GetAsciiStringAsLong(string str)
-        }
-
-        private static unsafe string GetAsciiStringImplementation(char* output, byte[] input, int inputOffset, int length)
-        {
-            for (var i = 0; i < length; i++)
+            if (start.IsDefault || end.IsDefault)
             {
-                output[i] = (char)input[inputOffset + i];
+                return null;
             }
 
-            return new string(output, 0, length);
-        }
-
-        private static unsafe string GetAsciiStringStack(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            // avoid declaring other local vars, or doing work with stackalloc
-            // to prevent the .locals init cil flag , see: https://github.com/dotnet/coreclr/issues/1279
-            char* output = stackalloc char[length];
-
-            return GetAsciiStringImplementation(output, start, end, inputOffset, length);
-        }
-
-        private unsafe static string GetAsciiStringHeap(MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            var buffer = new char[length];
+            var length = start.GetLength(end);
 
-            fixed (char* output = buffer)
+            if (length == 0)
             {
-                return GetAsciiStringImplementation(output, start, end, inputOffset, length);
+                return null;
             }
-        }
 
-        private static unsafe string GetAsciiStringImplementation(char* output, MemoryPoolBlock2 start, MemoryPoolIterator2 end, int inputOffset, int length)
-        {
-            var outputOffset = 0;
-            var block = start;
-            var remaining = length;
+            // Bytes out of the range of ascii are treated as "opaque data" 
+            // and kept in string as a char value that casts to same input byte value
+            // https://tools.ietf.org/html/rfc7230#section-3.2.4
 
-            var endBlock = end.Block;
-            var endIndex = end.Index;
+            var inputOffset = start.Index;
+            var block = start.Block;
 
-            while (true)
+            var asciiString = new string('\0', length);
+
+            fixed (char* outputStart = asciiString)
             {
-                int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+                var output = outputStart;
+                var remaining = length;
 
-                if (following > 0)
+                var endBlock = end.Block;
+                var endIndex = end.Index;
+
+                while (true)
                 {
-                    var input = block.Array;
-                    for (var i = 0; i < following; i++)
+                    int following = (block != endBlock ? block.End : endIndex) - inputOffset;
+
+                    if (following > 0)
                     {
-                        output[i + outputOffset] = (char)input[i + inputOffset];
+                        fixed (byte* blockStart = block.Array)
```

Comment:
There is a further combining change that will this fixed block (as buffers are already fixed) - but this stands alone as is.